go/tasks/plugins/k8s/ray/ray.go

```diff
@@ -0,0 +1,353 @@
+func (rayJobResourceHandler) BuildResource(ctx context.Context, taskCtx pluginsCore.TaskExecutionContext) (client.Object, error) {
+	taskTemplate, err := taskCtx.TaskReader().Read(ctx)
+	if err != nil {
+		return nil, errors.Errorf(errors.BadTaskSpecification, "unable to fetch task specification [%v]", err.Error())
+	} else if taskTemplate == nil {
+		return nil, errors.Errorf(errors.BadTaskSpecification, "nil task specification")
+	}
+
+	rayJob := plugins.RayJob{}
+	err = utils.UnmarshalStruct(taskTemplate.GetCustom(), &rayJob)
+	if err != nil {
+		return nil, errors.Errorf(errors.BadTaskSpecification, "invalid TaskSpecification [%v], Err: [%v]", taskTemplate.GetCustom(), err.Error())
+	}
+
+	podSpec, err := flytek8s.ToK8sPodSpec(ctx, taskCtx)
+	if err != nil {
+		return nil, errors.Errorf(errors.BadTaskSpecification, "Unable to create pod spec: [%v]", err.Error())
+	}
+
+	headReplicas := int32(1)
+	headNodeRayStartParams := make(map[string]string)
+	if rayJob.RayCluster.HeadGroupSpec != nil && rayJob.RayCluster.HeadGroupSpec.RayStartParams != nil {
+		headNodeRayStartParams = rayJob.RayCluster.HeadGroupSpec.RayStartParams
+	}
+	headNodeRayStartParams["include-dashboard"] = "true"
+	headNodeRayStartParams["node-ip-address"] = "$MY_POD_IP"
+	headNodeRayStartParams["dashboard-host"] = "0.0.0.0"
+
+	enableIngress := true
+	rayClusterSpec := rayv1alpha1.RayClusterSpec{
+		HeadGroupSpec: rayv1alpha1.HeadGroupSpec{
+			Template:       buildHeadPodTemplate(podSpec),
+			ServiceType:    v1.ServiceType(GetConfig().ServiceType),
+			Replicas:       &headReplicas,
+			EnableIngress:  &enableIngress,
+			RayStartParams: headNodeRayStartParams,
+		},
+		WorkerGroupSpecs: []rayv1alpha1.WorkerGroupSpec{},
+	}
+
+	for _, spec := range rayJob.RayCluster.WorkerGroupSpec {
+		workerPodTemplate := buildWorkerPodTemplate(podSpec)
+
+		minReplicas := spec.Replicas
+		maxReplicas := spec.Replicas
+		if spec.MinReplicas != 0 {
+			minReplicas = spec.MinReplicas
+		}
+		if spec.MaxReplicas != 0 {
+			maxReplicas = spec.MaxReplicas
+		}
+
+		workerNodeRayStartParams := make(map[string]string)
+		if spec.RayStartParams != nil {
+			workerNodeRayStartParams = spec.RayStartParams
+		}
+		workerNodeRayStartParams["node-ip-address"] = "$MY_POD_IP"
+
+		workerNodeSpec := rayv1alpha1.WorkerGroupSpec{
+			GroupName:      spec.GroupName,
+			MinReplicas:    &minReplicas,
+			MaxReplicas:    &maxReplicas,
+			Replicas:       &spec.Replicas,
+			RayStartParams: workerNodeRayStartParams,
+			Template:       workerPodTemplate,
+		}
+
+		rayClusterSpec.WorkerGroupSpecs = append(rayClusterSpec.WorkerGroupSpecs, workerNodeSpec)
+	}
+
+	serviceAccountName := flytek8s.GetServiceAccountNameFromTaskExecutionMetadata(taskCtx.TaskExecutionMetadata())
+
+	rayClusterSpec.HeadGroupSpec.Template.Spec.ServiceAccountName = serviceAccountName
+	for index := range rayClusterSpec.WorkerGroupSpecs {
+		rayClusterSpec.WorkerGroupSpecs[index].Template.Spec.ServiceAccountName = serviceAccountName
```

Comment:
is there ever a use-case where different worker groups have different service accounts? @daikeshi